src/internal.c

```diff
@@ -6005,30 +6078,35 @@ int HashOutputRaw(WOLFSSL* ssl, const byte* output, int sz)
         ssl->fuzzerCb(ssl, output, sz, FUZZ_HASH, ssl->fuzzerCtx);
 #endif
 #ifndef NO_OLD_TLS
-#ifndef NO_SHA
-    wc_ShaUpdate(&ssl->hsHashes->hashSha, output, sz);
-#endif
-#ifndef NO_MD5
-    wc_Md5Update(&ssl->hsHashes->hashMd5, output, sz);
-#endif
+    #ifndef NO_SHA
+        wc_ShaUpdate(&ssl->hsHashes->hashSha, output, sz);
+    #endif
+    #ifndef NO_MD5
+        wc_Md5Update(&ssl->hsHashes->hashMd5, output, sz);
+    #endif
 #endif /* NO_OLD_TLS */
 
     if (IsAtLeastTLSv1_2(ssl)) {
-#ifndef NO_SHA256
+    #ifndef NO_SHA256
         ret = wc_Sha256Update(&ssl->hsHashes->hashSha256, output, sz);
         if (ret != 0)
             return ret;
-#endif
-#ifdef WOLFSSL_SHA384
+    #endif
+    #ifdef WOLFSSL_SHA384
         ret = wc_Sha384Update(&ssl->hsHashes->hashSha384, output, sz);
         if (ret != 0)
             return ret;
-#endif
-#ifdef WOLFSSL_SHA512
+    #endif
+    #ifdef WOLFSSL_SHA512
         ret = wc_Sha512Update(&ssl->hsHashes->hashSha512, output, sz);
         if (ret != 0)
             return ret;
-#endif
+    #endif
+    #if defined(HAVE_ED25519) && !defined(WOLFSSL_NO_CLIENT_AUTH)
+        ret = Ed25519Update(ssl, output, sz);
```

Comment:
This seems like a huge waste of resources.  Someone turns on ed25519, maybe only for use at the wolfCrypt level for an occasional update verify or something but now they have to cache every handshake message.  Is there a better way to handle this?

Reply:
Nope. Ed25519 signatures hash some data then the messages then in a new hash, hash the output and the messages.
Really bad for embedded to have to store all messages.
I tried my best to find ways to turn message storing code off until explicitly needed. Only other option is to have a define for ED25519 in TLS 1.2.
TLS 1.3 uses the hash of the messages and not the messages themselves in the signature.

Reply:
We know when a user does not load a private key for TLS, or easily can.  We know the type of key, or easily can.  We also know the sig algo types after server hello.  But you're saying when a client does NOT even load a private key, lot alone an ed25519 key, he has to pay for caching 100% of the handshake messages.  I feel like we can be smarter here.

Reply:
The messages only need to be kept for Client Auth.
I can tell if the client has an ed25519 key and only cache messages when this happens.
On the server side, it will say that it supports all algorithms that are compiled in. Peer verification only happens when !resuming and verifyPeer is set. I now cache on these two conditions.